### PR TITLE
feature/contact-regex

### PR DESCRIPTION
**Pull request by xaverfleer**
This PR is a short version of PR #1. It is the same implementation with a new (shorter) branch name.

From #1:
```
This PR solves [d9e7646#r32237710](https://github.com/raphaelschaad/Plus-Web/commit/d9e7646a77454df3823caa50927af0bb004420a6#r32237710): Currently, the `contact` input field doesn't have a client-side validation other than `required`. This PR uses the HTML5 `pattern` attribute to add a REGEX that only allows valid emails and international mobile phone numbers.

Entries that should pass:
    * 077 44 66 505
    * 0774466505
    * 077/44 66 505
    * 077`44`66`505
    * 077'44'66'505
    * 077-44-66-505
    * (077)44 66 505
    * 0041774466505
    * +41774466505
    * [xaver.fleer@gmail.com](mailto:xaver.fleer@gmail.com)
    * [xaver.fleer+Plus@gmail.com](mailto:xaver.fleer+Plus@gmail.com)
    * xaver-fleer_89%[_@gmail.com](mailto:_@gmail.com)

Entries that should not pass
    * 077
    * xaver-fleerAtgmail.com
    * xaver-fleer@gmail
```

### CODE REVIEW

File: web/zfbc/index.html [hunk 1]
```diff
@@ -121,9 +134,15 @@ <h3>Weitere Inhalte demnächst verfügbar</h3>
...
+        <p id="contact">
+          Kontaktieren Sie uns unter
+          <a href="mailto:hello@plus.healthcare" target="_blank" rel="nofollow">hello@plus.healthcare</a>
+          oder
+          <a href="tel:+41786465577" rel="nofollow" class="nobr">+41&#8239;78&#8201;646&#8201;55&#8201;77</a>
+        </p>
         <p id="copyright">©&nbsp;2019 Plus Healthcare</p>
       </div>
     </footer>
+    <script src='main.js'></script>
```
Comment (raphaelschaad) [inline web/zfbc/index.html]: The PR wouldn't actually add a working feature once merged 😢
```suggestion
    <script src='./../main.js'></script>
```

File: web/zfbc/index.html [hunk 2]
```diff
@@ -67,13 +67,26 @@ <h4>Aufklärung zu möglichen Komplikationen bei einer Magenbypass-Operation</h4
...
+                <br>
+                <input type="date" name="birthday" size="20" maxlength="100" autocomplete="off" title="Bitte geben Sie das Geburtsdatum ein." required>
+              </label>
             </p>
             <p>
-              <label>E-Mail oder Mobiltelefon<br><input type="text" name="contact" size="30" maxlength="320" autocomplete="off" title="Bitte geben Sie eine gültige E-Mail-Adresse oder Mobiltelefonnummer ein." required></label>
+              <label>
+                E-Mail oder Mobiltelefon
+                <br>
+                <!-- pattern in contact field is added by javascript. Keep name attribute and the css selector in sync. -->
```
Comment (raphaelschaad) [inline web/zfbc/index.html]: Minor feedback: semantically this comment should have been committed with f9fb93f5fa6cdfd1d05e870f1cf31d9e05a14b03. For me the visually easy GitHub Desktop.app works best to stage/commit my work.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "plus-web",
+  "version": "0.0.1",
```
Comment (raphaelschaad) [inline package.json]: Should `"private": true`?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    "type": "git",
+    "url": "git+https://github.com/raphaelschaad/Plus-Web.git"
+  },
+  "author": "Raphael Schaad",
+  "license": "UNLICENSED",
+  "bugs": {
+    "url": "https://github.com/raphaelschaad/Plus-Web/issues"
+  },
+  "homepage": "https://github.com/raphaelschaad/Plus-Web#readme",
+  "dependencies": {
```
Comment (raphaelschaad) [inline package.json]: Is testing a `dependencies` or a `devDependencies`?

File: web/zfbc/index.html (hunk 1 above)
Comment (xaverfleer) [reply web/zfbc/index.html]: oh, you found it before me
2a8bd35 fixes it

File: web/zfbc/index.html (hunk 2 above)
Comment (xaverfleer) [reply web/zfbc/index.html]: I agree

Comment (xaverfleer) [reply web/zfbc/index.html]: added github desktop app to my to-have-a-look list

File: package.json (hunk 4 above)
Comment (xaverfleer) [reply package.json]: devdependencies

Comment (xaverfleer) [reply package.json]: 908b2a0 fixes this

File: package.json (hunk 3 above)
Comment (xaverfleer) [reply package.json]: not sure, should it? I do not yet understand the deploying process

**Comment by xaverfleer** [comment]
`assert.strictEqual(regex.test('0786465577x')` should not pass in the browser, if it does, there is an error..

**Comment by xaverfleer** [comment]
> Would it make sense for the unit tests to live in the root folder? I've treated `web/` as the "deployment folder"
No strong opinion. Weak opinion: I usually save *Test.js files next to their *.js. However I can live happily without this pattern.

**Comment by xaverfleer** [comment]
> (maybe we should rename it to `public/` at some point?)

good idea, this seems more common/familiar

**Comment by xaverfleer** [comment]
* Speaking of tape, have you compared it against

**Comment by xaverfleer** [comment]
* Speaking of tape, have you compared it against Mocha+Chai or Jest?

Not very deeply. I will definitly try Jest one day. Their snapshot mechanism looks promising.

 So far, I like tape because it's really simple -> simple to read and difficult to get distracted by doing complicated tests.

**Comment by xaverfleer** [comment]
* how to best run the repo (is npm run serve how you do it?)

for now, I'm just opening files in the browser. Compiling, adding a webserver and hot-reloading are a nice feature to add

**Comment by xaverfleer** [comment]
Thank you so much for your great inputs! great work!

File: package.json (hunk 3 above)
Comment (raphaelschaad) [reply package.json]: Seems to be a more sensible default to me: 

> If you set "private": true in your package.json, then npm will refuse to publish it.

https://docs.npmjs.com/files/package.json#private

Note: this has nothing to do with the Netlify deployment process and is just about adding an (npm) package.json to this repo.

File: web/zfbc/index.html [hunk 5]
```diff
@@ -143,6 +143,6 @@ <h3>Weitere Inhalte demnächst verfügbar</h3>
         <p id="copyright">©&nbsp;2019 Plus Healthcare</p>
       </div>
     </footer>
-    <script src='main.js'></script>
+    <script src='../main.js'></script>
```
Comment (raphaelschaad) [inline web/zfbc/index.html]: Mind following the convention of the surrounding code? I.e.: Using double quotes and the verbose path (`./../main.js`)? I think for HTML double quotes is pretty standard and I'd like to go with that convention. Ok for you? Using the verbose path I don't feel passionate about and we can discuss -- in either case, let's aim to be consistent.

**Comment by raphaelschaad** [comment]
Mind running npm update and updating the lock file? PS: The renaming of the `web/` folder doesn't have to happen on this branch, I can do that on master at some point in the future. Tracking it with fa4cf4a7e89927f8868a39a37be87056dbf92430

**Comment by raphaelschaad** [comment]
Is `a@a` really a valid email address? It passes when testing in the browser.

File: web/zfbc/index.html (hunk 1 above)
Comment (raphaelschaad) [reply web/zfbc/index.html]: Some companies have dedicated QA. I see QA not as a separate job but as a responsibility each Dev has for her features 💪 Always be testing your work *from an end user perspective* before pushing 👍

**Comment by xaverfleer** [comment]
> Is `a@a` really a valid email address? It passes when testing in the browser.

I never would have guessed, but it really is according to mdn: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email

**Comment by xaverfleer** [comment]
> Mind running npm update and updating the lock file? PS: The renaming of the `web/` folder doesn't have to happen on this branch, I can do that on master at some point in the future. Tracking it with [fa4cf4a](https://github.com/raphaelschaad/Plus-Web/commit/fa4cf4a7e89927f8868a39a37be87056dbf92430)

908b2a0 updates the dependencies

File: web/zfbc/index.html (hunk 5 above)
Comment (xaverfleer) [reply web/zfbc/index.html]: 95aa6ff fixes inconsistency

I agree 100%. Should we introduce a convention-set, for future developers? maybe this is the task of a html linter..

Comment (raphaelschaad) [reply web/zfbc/index.html]: Yes, a section in the README could be good. Something to do separately from this PR. Want to take a crack at it? By the way, even if it's just for the "future me" (me, you, anyone), I find it a good practice to write things down.

I've evaluated prettyhtml and found it way too opinionated (+bad defaults and no options). Not a high priority, but I'm open to suggestions.

File: mainTest.js [hunk 6]
```diff
@@ -58,5 +58,10 @@ test('contactRegex', (assert) => {
         assert.end();
     });
 
+    test('a@a should actually be a valid email address', (assert) => {
```
Comment (raphaelschaad) [inline mainTest.js]: Minor: "should actually be valid" according to what? the MDN link? would be great to have it/see it right there (for the future you) ;)

Comment (raphaelschaad) [reply mainTest.js]: After having slept over this, this is silly. Then we might as well just check `contains('@')` and call it a day. Let's use the "General Email Regex (RFC 5322 Official Standard)" that I've linked to in the very beginning. According to that Regex, a@a wouldn't pass. Let's document the quoted above where we cite the source.

Comment (raphaelschaad) [reply mainTest.js]: After having slept over this, this is silly. Then we might as well just check `contains('@')` and call it a day. Let's use the "General Email Regex (RFC 5322 Official Standard)" that I've linked to in the very beginning. According to that Regex, a@a wouldn't pass. Let's document the quoted above where we cite the source.

**Comment by xaverfleer** [comment]
tl;dr: email are opinionated. I prefer HTML 5.1. Either way, I do not expect any problem.

hmm.. not sure if this is a good idea. Email addresses seem to be an opinionated topic.

according to [emailregex](http://emailregex.com/), [this](https://3gtdu82m1uuz1tgbjb6a3704-wpengine.netdna-ssl.com/wp-content/uploads/sites/2/2014/06/General-Email-Regex-Railroad-Diagram-emailregex.com_.png) visual representation shows  the RFC 5322 definition.

that means, this is a valid email address: /@[0.0]

According to [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email), W3C defines [HTML5.1](https://www.w3.org/TR/html51/sec-forms.html#email-state-typeemail)

W3c stresses: "This requirement is a willful violation of RFC 5322, which defines a syntax for e-mail addresses that is simultaneously too strict (before the "@" character), too vague (after the "@" character), and too lax (allowing comments, whitespace characters, and quoted strings in manners unfamiliar to most users) to be of practical use here."

If you prefer, we can go with RFC 5322. I'd prefer to stick to HTML 5.1 (what we use now) because it's how browsers implement it by default.

Either way, I dont think we'll ever encounter any serious problem with mischecked email addresses.

**Comment by raphaelschaad** [comment]
Ok, let's do it the way you suggest. (In the end, we'll have to server-side validate whether the delivery ended up successful or not anyway and notify async on success [in the activity list] or on error [through a more intrusive notification]).

1. Let's use the regex

**Comment by raphaelschaad** [comment]
Ok, fair argument, let's do it the way you suggest:

1. Let's use exactly the following regex: `/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/`
2. Let's write a clear comment with the source (`https://www.w3.org/TR/html51/sec-forms.html#valid-e-mail-address`) and reasoning (part of your argument, part from the linked website above, and [part from Wikipedia](https://en.wikipedia.org/wiki/Email_address), something along the lines of: This regex follows RFC 1123 and willfully violates the addr-spec defined in RFC 5322 according to ... because it may contain elements that are not part of addresses commonly used ...).
3. To fully implement the recommended spec, do we also have to run the **value sanitization algorithm** as described in the link you provided? If yes, let's do it. If no, let's document it.

Eventually we'll have to validate the actual delivery server-side anyway and expose the result in an async fashion to the user. On success, we'll show the entry under activity, and on failure, we'll additionally show a more intrusive notification to get the user's attention that an action is needed.

**Comment by xaverfleer** [comment]
>     3\. value sanitization algorithm as described in the link you provided?

hmm.. can't find it the algorithm. what link do you mean?

**Comment by xaverfleer** [comment]
> Eventually we'll have to validate the actual delivery server-side anyway and expose the result in an async fashion to the user. On success, we'll show the entry under activity, and on failure, we'll additionally show a more intrusive notification to get the user's attention that an action is needed.

agree, this is something for the backlog

**Comment by raphaelschaad** [comment]
> The [value sanitization algorithm](https://www.w3.org/TR/html51/sec-forms.html#value-sanitization-algorithm) is as follows: Strip line breaks from the value, then strip leading and trailing whitespace from the value.

> Constraint validation: While the value of the element is neither the empty string nor a single valid e-mail address, the element is suffering from a type mismatch.

Source (your link): https://www.w3.org/TR/html51/sec-forms.html#element-statedef-input-e-mail

File: web/main.js [hunk 7]
```diff
@@ -7,8 +7,10 @@
 //  - number characters can be separated by special characters: (, ), <whitespaces> ` - / '
 // example that phoneRegex matches: (+41) 77 / 123 '45 '67
 const phoneRegex = '\^\\+?(?:[()\\s\'`\\-\\/]*\\d){10,15}\\s*$';
-// source of emailRegex: https://emailregex.com/
-const emailRegex = '^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$';
+
+// source of emailRegex: https://www.w3.org/TR/html51/sec-forms.html#valid-e-mail-address
+// This regex follows RFC 1123 and willfully violates the addr-spec defined in RFC 5322 according to https://www.w3.org/TR/html51/sec-forms.html#valid-e-mail-address because it may contain elements that are not part of addresses commonly used. We follow this standard, because this is what common browsers implement.
+const emailRegex = '^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$';
```
Comment (raphaelschaad) [inline web/main.js]: Source link, reasoning comment, regex implementation, all LGTM!

**Comment by xaverfleer** [comment]
> all LGTM!

not sure what you mean.. 🤔 [urban dictionary](https://www.urbandictionary.com/define.php?term=LGTM) proposes different meanings
- Let's Get This Merged
- Literally Garbage To Me

which one did you mean?

**Comment by raphaelschaad** [comment]
Hah, why ignore [the top definition](https://www.urbandictionary.com/define.php?term=LGTM)?

**Comment by xaverfleer** [comment]
=) alright

there is one last thing I'd like to implement: value sanitization. for now " xaver.fleer@gmail.com " is not valid (this is a bug IMO)

**Comment by raphaelschaad** [comment]
Agreed. Re-request a review in the sidebar when you're ready.

**Comment by xaverfleer** [comment]
this is a request to review 8f1c277

**Comment by raphaelschaad** [comment]
FYI I think the sidebar provides "native" features to manage the review lifecycle:

image

File: web/main.js [hunk 8]
```diff
@@ -10,7 +10,8 @@ const phoneRegex = '\^\\+?(?:[()\\s\'`\\-\\/]*\\d){10,15}\\s*$';
 
 // source of emailRegex: https://www.w3.org/TR/html51/sec-forms.html#valid-e-mail-address
 // This regex follows RFC 1123 and willfully violates the addr-spec defined in RFC 5322 according to https://www.w3.org/TR/html51/sec-forms.html#valid-e-mail-address because it may contain elements that are not part of addresses commonly used. We follow this standard, because this is what common browsers implement.
-const emailRegex = '^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$';
+// For value sanitization, leading and trailing whitespaces are allowed.
```
Comment (raphaelschaad) [inline web/main.js]: I like the practical implementation of the value sanitization "algorithm" by including it in the regex.

Are we concerned about the [strip line breaks](https://www.w3.org/TR/html51/infrastructure.html#stripped-line-breaks)-part of the description of the original algorithm?

File: mainTest.js [hunk 9]
```diff
@@ -63,5 +63,15 @@ test('contactRegex', (assert) => {
         assert.end();
     });
 
+    test('leading spaces are valid', (assert) => {
+        assert.strictEqual('   xaver@fleer.com'.match(regex)[0], '   xaver@fleer.com');
```
Comment (raphaelschaad) [inline mainTest.js]: Why do we use a different syntax than above (`regex.test(...)`)?

File: mainTest.js [hunk 10]
```diff
@@ -63,5 +63,15 @@ test('contactRegex', (assert) => {
...
+        assert.end();
+    });
+
+    test('trailing spaces are valid', (assert) => {
+        assert.strictEqual('xaver@fleer.com  '.match(regex)[0], 'xaver@fleer.com  ');
+        assert.end();
+    });
+
     assert.end();
 })
```
Comment (raphaelschaad) [inline mainTest.js]: Can we add a test with multiple email addresses? E.g.: 'eva@fleer.com,xaver@fleer.com`

It should be INVALID given "the multiple attribute is not specified on the element" ([source](https://www.w3.org/TR/html51/sec-forms.html#email-state-typeemail))

**Comment by xaverfleer** [comment]
Thanks for indicating how to re-request a review. I find it surprisingly difficult to use github's features. In this case, the icon is not intoitive to me + the hover effect does not work in firefox.

anyway, can you explain me, how to start conversations as you do all the time (e.g. the one just obove this comment)?

**Comment by xaverfleer** [comment]
about the html lint discussion above: is there a way to add another comment? (I could not find one)

It seems like [htmllint](https://www.npmjs.com/package/htmllint) is a nice linter. it's the one with the most weekly downloads I could find. I've checked it out; it show quite some interesting information about (our) html. I agree to tackle this topic in a new PR.

**Comment by xaverfleer** [comment]
i now found the conversation feature. no more need for instructions about this 👍

File: web/main.js (hunk 8 above)
Comment (raphaelschaad) [reply web/main.js]: Sorry to unresolve this, but what about the open question?!

**Comment by xaverfleer** [comment]
db99e0d and 0d7f7fc rebase the branch onto the master.

To resolve the merge conflict, I undid html linting which IMO belongs to a separate story.

about the rebase: I was surprised how few changes were actually made during this whole process (+/- 25 lines code + comments + tests)

about the process: I agree in both points (long-windedness and learning experience). It shaped my mind - mainly about commits - in a good way.

about the result: great quality!

Well done 🙌

**Comment by raphaelschaad** [comment]
In the "Files changed" tab of any PR (all the way at the top) you can always see the absolute diff. I like to keep track of the actual changeset when working on a PR and not let the historical commit noise cloud my perception of what is actually being landed on master.

I *love* how you got it down to two semantical clean commits. Now, while this is what we should be aiming for each time we open a PR, it's not realistic to get everything right from the beginning nor should a false sense of perfectionism hold us back from sharing our work early and submitting it for review. There's a balance and as you get more experienced your PRs will require less feedbacks or the discussions will more center around higher-level architecture vs. implementation details.

So, important elements we now have:
- semantical, short branch name/title
- description with thinking that went into the PR
- clear call to action for review
- code tested by the submitter that doesn't break master
- code tested by the submitter that does what it says
- tests (often not needed)
- no merge conflicts
- nothing lumped into the PR that doesn't belong there (formatting of other areas)
- clean up (e.g. todo removed)

Now, hit that green merge button ✅SHIP IT! 🚢

**Comment by xaverfleer** [comment]
Thank you for telling me about the "Files changed"-feature
 + nice post mortem list.

**Comment by xaverfleer** [comment]
i could not find "that green merge button" so I merged in the CLI. Which tool do you usually use?

**Comment by raphaelschaad** [comment]
Sure! Explore GH a bit, maybe read a tutorial and set up a playground repo for yourself.

Are you able to merge this PR? I think you should have the necessary rights to do it, no?

PS: master has since moved and there are some new (easy) conflicts to be resolved.

**Comment by raphaelschaad** [comment]
I just use GH. You won't see a merge button if there are conflicts. Master has moved since I've written that you should go ahead and merge. I'm pretty sure that if you had resolved the merge conflicts and pushed to this branch, you would have seen the merge button just now. If you continue to have difficulties with this, let me know. It's a standard flow.

https://help.github.com/en/articles/merging-a-pull-request

Explore GH a bit, maybe read a tutorial and set up a repo for yourself as a playground to do dry runs through these motions.

**Comment by xaverfleer** [comment]
found it, thank you
